### PR TITLE
Adds the content-addressable, deletable ElementsAttr scaffolding.

### PR DESCRIPTION
**Pull request by stellaraccident**
This is a preview for the moment and has work needed to make it complete:

* [] CasManagerDialectInterface needs to be lock protected, since it can be used by multiple threads.
* [] Parser hooks need to register blobs with the CasManagerDialectInterface.
* [] Invalidation needs to allow a retain set.
* [] Implementation/tests for ElementsAttr accessors.
* [] Builder sugar to help us port some of the existing DenseElementsAttr transformations.

Sending now for early feedback since this is a substantial addition.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Util/IR/UtilInterfaces.td [hunk 1]
```diff
@@ -1081,6 +1081,53 @@ def Util_SubrangeOperandOpInterface : OpInterface<"SubrangeOperandOpInterface">
...
+      }],
+      /*retTy=*/"ElementsAttr",
+      /*methodName=*/"getElements",
+      /*args=*/(ins)
+    >,
+    InterfaceMethod<
+      /*desc=*/[{
+        Returns a new attribute which (to the extent possible) shares storage
+        but with a specified, compatible shape. The new type must have the
+        same total number of alements as well as element type.
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/IR/UtilInterfaces.td]:
```suggestion
        same total number of elements as well as element type.
```

File: compiler/src/iree/compiler/Dialect/Util/IR/UtilDialect.cpp [hunk 2]
```diff
@@ -28,23 +31,120 @@ namespace iree_compiler {
...
   }
+
+  std::string
+  getResourceKey(const AsmDialectResourceHandle &handle) const override {
+    return cast<UtilDialect::CasResourceHandle>(handle).getKey().str();
+  }
+
+  FailureOr<AsmDialectResourceHandle>
+  declareResource(StringRef key) const final {
+    llvm::errs() << "::: declareResource " << key << "\n";
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/IR/UtilDialect.cpp:115]: (draft) note to remove

File: compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td [hunk 3]
```diff
@@ -104,4 +105,87 @@ def Util_CompositeAttr : AttrDef<Util_Dialect, "Composite", [
...
+
+def Util_CasScope : AttrDef<Util_Dialect, "CasScope"> {
+  let mnemonic = "cas-scope";
+  let summary = "A unique scope delimitter for bindings CAS resources";
+  let description = [{
+    CAS resources can be globally (context bound) or locally (sub-tree)
+    bound. Locally scoped resources are eligible for various lifetime
+    policies, liveness analysis, etc, and they can be turned into tombstones
+    if it is deemed valid to do so (releasing the memory they hold while 
+    keeping them in index so an to not produce dangling pointers).
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td]:
```suggestion
    keeping them in index so as to not produce dangling pointers).
```

File: compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td [hunk 4]
```diff
@@ -104,4 +105,87 @@ def Util_CompositeAttr : AttrDef<Util_Dialect, "Composite", [
   let hasCustomAssemblyFormat = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// Resources
+//===----------------------------------------------------------------------===//
+
+def Util_CasScope : AttrDef<Util_Dialect, "CasScope"> {
+  let mnemonic = "cas-scope";
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td]: prefer `_` (we don't use `-` in any other IREE attributes - don't want to mix)

File: compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td [hunk 5]
```diff
@@ -104,4 +105,87 @@ def Util_CompositeAttr : AttrDef<Util_Dialect, "Composite", [
...
+  let summary = "A unique scope delimitter for bindings CAS resources";
+  let description = [{
+    CAS resources can be globally (context bound) or locally (sub-tree)
+    bound. Locally scoped resources are eligible for various lifetime
+    policies, liveness analysis, etc, and they can be turned into tombstones
+    if it is deemed valid to do so (releasing the memory they hold while 
+    keeping them in index so an to not produce dangling pointers).
+
+    A scope is demarcated by a #util.cas-scope attribute, which internally
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td:123]: what does cloning do here? if there was a func.func with a cas scope and a pattern cloned it or created a new one and setDialectAttrs(getDialectAttrs) to preserve the scope attr would that create a new scope or reuse the old one? what would you want to happen? (if we start scoping at the sub-module level we will have this happen, so it'd be good to be clear about that here)

**Review by benvanik — COMMENTED** [review]
s/Cas/CAS/g but is looking good!

File: compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td (hunk 5 above)
Comment (stellaraccident) [reply compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.td:123]: Live references to the attribute have pointer equivalence, so clone will reuse the existing scope.

File: compiler/src/iree/compiler/Dialect/Util/IR/UtilInterfaces.td (hunk 1 above)
Comment (stellaraccident) [reply compiler/src/iree/compiler/Dialect/Util/IR/UtilInterfaces.td]: I think this whole interface is from a prior draft. Commented out for now as I can't remember why I left it.

**Comment by benvanik** [comment]
@stellaraccident I believe this is still desirable but maybe parameters made it less pressing a need?

**Comment by benvanik** [comment]
(still WIP; renaming to users/stellaraccident/managed_resource)
```
  git branch -m managed_resource users/stellaraccident/managed_resource
  git fetch origin
  git branch -u origin/users/stellaraccident/managed_resource users/stellaraccident/managed_resource
  git remote set-head origin -a
```

**Pull request closed without merging**